<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>nbaR services examples • rOpenSci: nbaR</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="/favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="/favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="/apple-touch-icon.png">
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- docsearch --><script src="../docsearch.js"></script><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/docsearch.js/2.6.3/docsearch.min.css" integrity="sha256-QOSRU/ra9ActyXkIBbiIB144aDBdtvXBcNc3OTNuX/Q=" crossorigin="anonymous">
<link href="../docsearch.css" rel="stylesheet">
<script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/jquery.mark.min.js" integrity="sha256-4HLtjeVgH0eIB3aZ9mLYF6E8oU5chNdjU6p6rrXpl9U=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.11.1/css/all.min.css" integrity="sha256-PbSmjxuVAzJ6FPvNYsrXygfGhNJYyZ2GktDbkMBqQZg=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.11.1/css/v4-shims.min.css" integrity="sha256-A6jcAdwFD48VMjlI3GDxUd+eCQa7/KWy6G9oe/ovaPA=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/headroom.min.js" integrity="sha256-DJFC1kqIhelURkuza0AvYal5RxMtpzLjFhsnVIeuk+U=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><!-- ropensci --><link href="../ropensci.css" rel="stylesheet">
<meta property="og:title" content="nbaR services examples">
<meta property="og:description" content="nbaR">
<meta name="twitter:card" content="summary">
<!-- Matomo --><link rel="stylesheet" type="text/css" href="https://cdn.jsdelivr.net/npm/cookieconsent@3/build/cookieconsent.min.css">
<script src="https://cdn.jsdelivr.net/npm/cookieconsent@3/build/cookieconsent.min.js" data-cfasync="false"></script><script src="https://ropensci.org/scripts/matomo.js"></script><noscript><p><img src="https://ropensci.matomo.cloud/matomo.php?idsite=1&amp;rec=1" style="border:0;" alt=""></p></noscript>
<!-- End Matomo Code -->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>

      <div class="navbar-brand-container">
        <a href="https://ropensci.org"><img src="https://ropensci.org/img/icon_short_white.svg" id="hexlogo" alt="rOpenSci"></a>
        <a class="navbar-brand" href="../index.html"><i>nbaR <small>v0.1.0</small></i></a>
      </div>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav navbar-right">
<li>
  <a href="../index.html">
    <span class="fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../articles/nbaR.html">Get started</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Documentation
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/apiclient.html">Objects and advanced queries</a>
    </li>
    <li>
      <a href="../articles/services.html">Services summary</a>
    </li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Example Workflows
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/sharks.html">Calibrating a molecular phylogeny of the sharks</a>
    </li>
    <li>
      <a href="../articles/tomato.html">The oldest tomato specimen in the world</a>
    </li>
  </ul>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/apiclient.html">nbaR API client: concepts and object model</a>
    </li>
    <li>
      <a href="../articles/services.html">nbaR services examples</a>
    </li>
    <li>
      <a href="../articles/sharks.html">Calibrating a molecular phylogeny of the sharks</a>
    </li>
    <li>
      <a href="../articles/tomato.html">The oldest tomato specimen in the world</a>
    </li>
  </ul>
</li>
        <li>
  <a href="https://github.com/naturalis/nbaR">
    <span class="fas fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
<form class="navbar-form navbar-right docsearch" role="search">
        <div class="form-group">
          <input type="search" class="form-control" name="search-input" id="search-input" placeholder="Search..." aria-label="Search for..." autocomplete="off">
</div>
      </form>
      
    </div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>nbaR services examples</h1>
            
            <h4 class="date">Mar 13, 2021</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/ropensci/nbaR/blob/master/vignettes/services.Rmd"><code>vignettes/services.Rmd</code></a></small>
      <div class="hidden name"><code>services.Rmd</code></div>

    </div>

    
    
<div id="services-examples" class="section level1">
<h1 class="hasAnchor">
<a href="#services-examples" class="anchor"></a>Services examples</h1>
<div id="specimen-occurrence-services" class="section level2 tabset">
<h2 class="hasAnchor">
<a href="#specimen-occurrence-services" class="anchor"></a>Specimen occurrence services</h2>
<p>Specimen records constitute the core of data served by the NBA. Museum specimens can represent a whole variety of different objects such as plants, animals or single parts thereof, DNA samples, fossils, rocks or meteorites. For detailed information of the data model, please refer to the <a href="http://docs.biodiversitydata.nl/en/latest/doc-spec-services/specimen/">official documentation in the NBA</a>.</p>
<p>All specimen occurrence services are accessible using the methods within the <code>SpecimenClient</code> class. For a list of available endpoints, please refer to the class documentation (<code><a href="../reference/Specimen.html">?Specimen</a></code>). Below, we will give details about all services, grouped by category.</p>
<div id="query-services" class="section level3">
<h3 class="hasAnchor">
<a href="#query-services" class="anchor"></a>Query services</h3>
<div id="query" class="section level4">
<h4 class="hasAnchor">
<a href="#query" class="anchor"></a>query</h4>
<p>Querying for specimens is accomplished with the <code>query</code> method in the class <code>SpecimenClient</code>. For simple queries, query parameters of type <code>list</code> can be passed via the parameter <code>queryParams</code>, for example we can query specimens of the Family <em>Ebenaceae</em> that were collected in Europe:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="st"><a href="https://docs.ropensci.org/nbaR">'nbaR'</a></span><span class="op">)</span>

<span class="co"># instantiate specimen client</span>
<span class="va">sc</span> <span class="op">&lt;-</span> <span class="va"><a href="../reference/SpecimenClient.html">SpecimenClient</a></span><span class="op">$</span><span class="fu">new</span><span class="op">(</span><span class="op">)</span>

<span class="co"># specify query params in named list</span>
<span class="va">qp</span> <span class="op">&lt;-</span>
  <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span>
    identifications.defaultClassification.family <span class="op">=</span> <span class="st">"Ebenaceae"</span>,
    gatheringEvent.continent <span class="op">=</span> <span class="st">"Europe"</span>
  <span class="op">)</span>

<span class="co"># query</span>
<span class="va">res</span> <span class="op">&lt;-</span> <span class="va">sc</span><span class="op">$</span><span class="fu">query</span><span class="op">(</span>queryParams <span class="op">=</span> <span class="va">qp</span><span class="op">)</span></code></pre></div>
<p>If we now want to know all the countries that the specimens were collected in, we can access the <code>Specimen</code> objects in <code>res$content$resultSet</code> as follows:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/lapply.html">sapply</a></span><span class="op">(</span><span class="va">res</span><span class="op">$</span><span class="va">content</span><span class="op">$</span><span class="va">resultSet</span>, <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span><span class="va">x</span><span class="op">$</span><span class="va">item</span><span class="op">$</span><span class="va">gatheringEvent</span><span class="op">$</span><span class="va">country</span><span class="op">)</span></code></pre></div>
<pre><code>##  [1] "Netherlands" "Austria"     "Italy"       "Netherlands" "France"     
##  [6] "Switzerland" "Switzerland" "Netherlands" "Netherlands" "Netherlands"</code></pre>
<p>Note that passing query parameters as a named list only allows for limited queries; the logical conjunction between parameters is for example always <code>AND</code>. More complex queries can be accomplished using the <code>QuerySpec</code> object:</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># get all specimens with genus name starting with 'Hydro'</span>
<span class="va">qc</span> <span class="op">&lt;-</span>
  <span class="va"><a href="../reference/QueryCondition.html">QueryCondition</a></span><span class="op">$</span><span class="fu">new</span><span class="op">(</span>field <span class="op">=</span> <span class="st">"identifications.defaultClassification.genus"</span>,
                     operator <span class="op">=</span> <span class="st">"STARTS_WITH"</span>,
                     value <span class="op">=</span> <span class="st">"Hydro"</span><span class="op">)</span>
<span class="va">qs</span> <span class="op">&lt;-</span> <span class="va"><a href="../reference/QuerySpec.html">QuerySpec</a></span><span class="op">$</span><span class="fu">new</span><span class="op">(</span>conditions <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span><span class="va">qc</span><span class="op">)</span><span class="op">)</span>
<span class="va">res</span> <span class="op">&lt;-</span> <span class="va">sc</span><span class="op">$</span><span class="fu">query</span><span class="op">(</span><span class="va">qs</span><span class="op">)</span></code></pre></div>
</div>
<div id="download_query" class="section level4">
<h4 class="hasAnchor">
<a href="#download_query" class="anchor"></a>download_query</h4>
<p>The <code>query</code> function is limited to retrieve 50000 specimen at once (this is determined in the parameter <code>index.max_result_window</code>, the value is retrievable using the <code>getSettings</code> method in the metadata section). In order to provide access for a larger amount of data, the <code>query_download</code> takes the same arguments as <code>query</code>, but download the data as a gzip stream under the hood. Unlike <code>query</code>, <code>query_download</code> returns a list of specimen objects instead of a <code>ResultSet</code>. Example:</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">## get the first 100000 specimen objects (not possible with query method)</span>
<span class="va">res</span> <span class="op">&lt;-</span> <span class="va">sc</span><span class="op">$</span><span class="fu">download_query</span><span class="op">(</span><span class="va"><a href="../reference/QuerySpec.html">QuerySpec</a></span><span class="op">$</span><span class="fu">new</span><span class="op">(</span>size<span class="op">=</span><span class="fl">100000</span><span class="op">)</span><span class="op">)</span></code></pre></div>
</div>
</div>
<div id="data-access-services" class="section level3">
<h3 class="hasAnchor">
<a href="#data-access-services" class="anchor"></a>Data access services</h3>
<p>Several access methods offer the convenient retrieval of specimens matching a certain identifier or being part of a certain collection. Below we give examples of how to use the currently implemented data access services for specimen records:</p>
<div id="access-to-collections" class="section level4">
<h4 class="hasAnchor">
<a href="#access-to-collections" class="anchor"></a>Access to collections</h4>
<p>Some of the specimens available via the NBA are categorised thematically into special collections, such as the Siebold-, Dubois- or Jongmans collection. The function <code>get_named_collection</code> lists all available special collections and the identifiers of the specimens within a collection can be queries with <code>get_ids_in_collection</code></p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">sc</span><span class="op">$</span><span class="fu">get_named_collections</span><span class="op">(</span><span class="op">)</span><span class="op">$</span><span class="va">content</span></code></pre></div>
<pre><code>## Warning in self$handleError(response): javax.ws.rs.NotFoundException:
## RESTEASY003210: Could not find resource for full path: http://
## api.biodiversitydata.nl/v2/specimen/getNamedCollections</code></pre>
<pre><code>## [1] "javax.ws.rs.NotFoundException: RESTEASY003210: Could not find resource for full path: http://api.biodiversitydata.nl/v2/specimen/getNamedCollections"</code></pre>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">sc</span><span class="op">$</span><span class="fu">get_ids_in_collection</span><span class="op">(</span><span class="st">"siebold"</span><span class="op">)</span><span class="op">$</span><span class="va">content</span></code></pre></div>
<pre><code>## Warning in self$handleError(response): javax.ws.rs.NotFoundException:
## RESTEASY003210: Could not find resource for full path: http://
## api.biodiversitydata.nl/v2/specimen/getIdsInCollection/siebold</code></pre>
<pre><code>## [1] "javax.ws.rs.NotFoundException: RESTEASY003210: Could not find resource for full path: http://api.biodiversitydata.nl/v2/specimen/getIdsInCollection/siebold"</code></pre>
</div>
<div id="count" class="section level4">
<h4 class="hasAnchor">
<a href="#count" class="anchor"></a>count</h4>
<p>For any given query (with <code>QuerySpec</code> or not), returns the count of matches instead of specimen objects:</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Example with QuerySpec:</span>
<span class="co"># how many specimens are there in the 'Botany' collection?</span>
<span class="va">qc</span> <span class="op">&lt;-</span> <span class="va"><a href="../reference/QueryCondition.html">QueryCondition</a></span><span class="op">$</span><span class="fu">new</span><span class="op">(</span>field<span class="op">=</span><span class="st">'collectionType'</span>, operator<span class="op">=</span><span class="st">'EQUALS'</span>, value<span class="op">=</span><span class="st">'Botany'</span><span class="op">)</span>
<span class="va">qs</span> <span class="op">&lt;-</span> <span class="va"><a href="../reference/QuerySpec.html">QuerySpec</a></span><span class="op">$</span><span class="fu">new</span><span class="op">(</span>conditions<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span><span class="va">qc</span><span class="op">)</span><span class="op">)</span>

<span class="co"># get the number of specimens</span>
<span class="va">sc</span><span class="op">$</span><span class="fu">count</span><span class="op">(</span><span class="va">qs</span><span class="op">)</span><span class="op">$</span><span class="va">content</span></code></pre></div>
<pre><code>## [1] 4946578</code></pre>
<p>Note that the count of matches for a given query is also returned by the <code>query</code> function. However, <code>count</code> is more lightweight as it returns an <code>integer</code> instead of a <code>ResultSet</code> containing <code>Specimen</code> objects.</p>
</div>
<div id="exists" class="section level4">
<h4 class="hasAnchor">
<a href="#exists" class="anchor"></a>exists</h4>
<p>Check if a record exists, based on its <em>unitID</em>:</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># use SpecimenClient instantiated above</span>
<span class="va">res</span> <span class="op">&lt;-</span> <span class="va">sc</span><span class="op">$</span><span class="fu">exists</span><span class="op">(</span><span class="st">'ZMA.INS.1255440'</span><span class="op">)</span>                                        

<span class="co"># content is boolean</span>
<span class="va">res</span><span class="op">$</span><span class="va">content</span></code></pre></div>
<pre><code>## [1] TRUE</code></pre>
</div>
<div id="find" class="section level4">
<h4 class="hasAnchor">
<a href="#find" class="anchor"></a>find</h4>
<p>Return a single specimen given its identifier (Note: the identifier of a specimen is different from the unitID, see also <a href="http://docs.biodiversitydata.nl/en/latest/doc-spec-services/specimen/#specids">here</a>):</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">id</span> <span class="op">&lt;-</span> <span class="st">"RMNH.MAM.17209.B@CRS"</span>
<span class="va">res</span> <span class="op">&lt;-</span> <span class="va">sc</span><span class="op">$</span><span class="fu">find</span><span class="op">(</span><span class="va">id</span><span class="op">)</span>

<span class="co"># content is single specimen object</span>
<span class="va">res</span><span class="op">$</span><span class="va">content</span></code></pre></div>
<pre><code>## &lt;Specimen&gt;
## Fields:
##  sourceSystem:   object of class &lt;SourceSystem&gt; 
##  sourceSystemId:  RMNH.MAM.17209.B 
##  recordURI:   
##  id:  RMNH.MAM.17209.B@CRS 
##  unitID:  RMNH.MAM.17209.B 
##  unitGUID:    https://data.biodiversitydata.nl/naturalis/specimen/RMNH.MAM.17209.B 
##  collectorsFieldNumber:   
##  assemblageID:    
##  sourceInstitutionID:     Naturalis Biodiversity Center 
##  sourceID:    CRS 
##  previousSourceID:    
##  owner:   Naturalis Biodiversity Center 
##  licenseType:     Copyright 
##  license:     CC0 1.0 
##  recordBasis:     PreservedSpecimen 
##  kindOfUnit:  skin 
##  collectionType:  Mammalia 
##  sex:     female 
##  phaseOrStage:    juvenile 
##  title:   RMNH.MAM.17209.b_RMNH_MAM_17209.b 
##  notes:   
##  preparationType:     mounted skin 
##  previousUnitsText:   
##  numberOfSpecimen:    1 
##  fromCaptivity:   FALSE 
##  objectPublic:    TRUE 
##  multiMediaPublic:    TRUE 
##  acquiredFrom:   object of class &lt;Agent&gt; 
##  gatheringEvent: object of class &lt;GatheringEvent&gt; 
##  identifications:    list of length 1 
##  associatedMultiMediaUris:   list of length 0 
##  theme:   
## Methods:
##  fromJSONString
##  toJSONString
##  fromList
##  toList
##  print</code></pre>
</div>
<div id="find_by_ids" class="section level4">
<h4 class="hasAnchor">
<a href="#find_by_ids" class="anchor"></a>find_by_ids</h4>
<p>Same as <code>find</code>, but takes multiple IDs:</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">ids</span> <span class="op">&lt;-</span> <span class="st">"RMNH.INS.657083@CRS,L.1589244@BRAHMS"</span>
<span class="va">res</span> <span class="op">&lt;-</span> <span class="va">sc</span><span class="op">$</span><span class="fu">find_by_ids</span><span class="op">(</span><span class="va">ids</span><span class="op">)</span></code></pre></div>
</div>
<div id="find_by_unit_id" class="section level4">
<h4 class="hasAnchor">
<a href="#find_by_unit_id" class="anchor"></a>find_by_unit_id</h4>
<p>Find a specimen by its <code>unitID</code>:</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">unitID</span> <span class="op">&lt;-</span> <span class="st">"RMNH.MAM.1513"</span>
<span class="va">res</span> <span class="op">&lt;-</span> <span class="va">sc</span><span class="op">$</span><span class="fu">find_by_unit_id</span><span class="op">(</span><span class="va">unitID</span><span class="op">)</span></code></pre></div>
</div>
</div>
<div id="aggregation-services" class="section level3">
<h3 class="hasAnchor">
<a href="#aggregation-services" class="anchor"></a>Aggregation services</h3>
<p>Aggregation services group available data according to different criteria.</p>
<div id="get_distinct_values" class="section level4">
<h4 class="hasAnchor">
<a href="#get_distinct_values" class="anchor"></a>get_distinct_values</h4>
<p>This method takes a specific field as an argument and returns all possible values and the frequency for that field in the data. Below we get all possible values for the country in which a specimen was collected.</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">sc</span><span class="op">$</span><span class="fu">get_distinct_values</span><span class="op">(</span><span class="st">"gatheringEvent.country"</span><span class="op">)</span></code></pre></div>
<p><strong>Note</strong>: By default, <code>get_distinct_values</code> lists only the first 10 hits. The above query thus does not reflect the distinct values in the hole dataset. This number can be increased with e.g. setting the <code>size</code> parameter in a <code>QuerySpec</code> object passed to the method.</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">sc</span><span class="op">$</span><span class="fu">get_distinct_values</span><span class="op">(</span><span class="st">"gatheringEvent.country"</span>,
                       querySpec <span class="op">=</span> <span class="va"><a href="../reference/QuerySpec.html">QuerySpec</a></span><span class="op">$</span><span class="fu">new</span><span class="op">(</span>size <span class="op">=</span> <span class="fl">10000</span><span class="op">)</span><span class="op">)</span></code></pre></div>
</div>
<div id="count_distinct_values" class="section level4">
<h4 class="hasAnchor">
<a href="#count_distinct_values" class="anchor"></a>count_distinct_values</h4>
<p>Instead of returning all different values for a given field, this method does a mere count:</p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">sc</span><span class="op">$</span><span class="fu">count_distinct_values</span><span class="op">(</span><span class="st">"gatheringEvent.country"</span><span class="op">)</span></code></pre></div>
</div>
<div id="get_distinct_values_per_group" class="section level4">
<h4 class="hasAnchor">
<a href="#get_distinct_values_per_group" class="anchor"></a>get_distinct_values_per_group</h4>
<p>Suppose you want to add another filter to the above query of retrieving all distinct values for a given field, such as splitting the results by their respective source system:</p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">sc</span><span class="op">$</span><span class="fu">get_distinct_values_per_group</span><span class="op">(</span><span class="st">"sourceSystem.name"</span>, <span class="st">"gatheringEvent.country"</span><span class="op">)</span></code></pre></div>
</div>
<div id="count_distinct_values_per_group" class="section level4">
<h4 class="hasAnchor">
<a href="#count_distinct_values_per_group" class="anchor"></a>count_distinct_values_per_group</h4>
<p>If no return of the actual values is needed, a simple count per group is done as follows:</p>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">sc</span><span class="op">$</span><span class="fu">count_distinct_values_per_group</span><span class="op">(</span><span class="st">"sourceSystem.name"</span>,
                                   <span class="st">"gatheringEvent.country"</span><span class="op">)</span><span class="op">$</span><span class="va">content</span></code></pre></div>
</div>
</div>
<div id="metadata-services" class="section level3">
<h3 class="hasAnchor">
<a href="#metadata-services" class="anchor"></a>Metadata services</h3>
<p>Specimen Metadata services include the same standard metadata services as for the other data types:</p>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># get all paths for the Specimen datatype </span>
<span class="va">sc</span><span class="op">$</span><span class="fu">get_paths</span><span class="op">(</span><span class="op">)</span><span class="op">$</span><span class="va">content</span></code></pre></div>
<pre><code>##   [1] "sourceSystem.code"                                                              
##   [2] "sourceSystem.name"                                                              
##   [3] "sourceSystemId"                                                                 
##   [4] "recordURI"                                                                      
##   [5] "unitID"                                                                         
##   [6] "unitGUID"                                                                       
##   [7] "collectorsFieldNumber"                                                          
##   [8] "assemblageID"                                                                   
##   [9] "sourceInstitutionID"                                                            
##  [10] "sourceID"                                                                       
##  [11] "previousSourceID"                                                               
##  [12] "owner"                                                                          
##  [13] "licenseType"                                                                    
##  [14] "license"                                                                        
##  [15] "recordBasis"                                                                    
##  [16] "kindOfUnit"                                                                     
##  [17] "collectionType"                                                                 
##  [18] "sex"                                                                            
##  [19] "phaseOrStage"                                                                   
##  [20] "title"                                                                          
##  [21] "notes"                                                                          
##  [22] "preparationType"                                                                
##  [23] "previousUnitsText"                                                              
##  [24] "numberOfSpecimen"                                                               
##  [25] "fromCaptivity"                                                                  
##  [26] "objectPublic"                                                                   
##  [27] "multiMediaPublic"                                                               
##  [28] "acquiredFrom.agentText"                                                         
##  [29] "gatheringEvent.projectTitle"                                                    
##  [30] "gatheringEvent.worldRegion"                                                     
##  [31] "gatheringEvent.continent"                                                       
##  [32] "gatheringEvent.country"                                                         
##  [33] "gatheringEvent.iso3166Code"                                                     
##  [34] "gatheringEvent.provinceState"                                                   
##  [35] "gatheringEvent.island"                                                          
##  [36] "gatheringEvent.locality"                                                        
##  [37] "gatheringEvent.city"                                                            
##  [38] "gatheringEvent.sublocality"                                                     
##  [39] "gatheringEvent.localityText"                                                    
##  [40] "gatheringEvent.dateTimeBegin"                                                   
##  [41] "gatheringEvent.dateTimeEnd"                                                     
##  [42] "gatheringEvent.dateText"                                                        
##  [43] "gatheringEvent.method"                                                          
##  [44] "gatheringEvent.altitude"                                                        
##  [45] "gatheringEvent.altitudeUnifOfMeasurement"                                       
##  [46] "gatheringEvent.behavior"                                                        
##  [47] "gatheringEvent.biotopeText"                                                     
##  [48] "gatheringEvent.depth"                                                           
##  [49] "gatheringEvent.depthUnitOfMeasurement"                                          
##  [50] "gatheringEvent.code"                                                            
##  [51] "gatheringEvent.establishmentMeans"                                              
##  [52] "gatheringEvent.gatheringPersons.agentText"                                      
##  [53] "gatheringEvent.gatheringPersons.fullName"                                       
##  [54] "gatheringEvent.gatheringPersons.organization.agentText"                         
##  [55] "gatheringEvent.gatheringPersons.organization.name"                              
##  [56] "gatheringEvent.gatheringOrganizations.agentText"                                
##  [57] "gatheringEvent.gatheringOrganizations.name"                                     
##  [58] "gatheringEvent.siteCoordinates.longitudeDecimal"                                
##  [59] "gatheringEvent.siteCoordinates.latitudeDecimal"                                 
##  [60] "gatheringEvent.siteCoordinates.gridCellSystem"                                  
##  [61] "gatheringEvent.siteCoordinates.gridLatitudeDecimal"                             
##  [62] "gatheringEvent.siteCoordinates.gridLongitudeDecimal"                            
##  [63] "gatheringEvent.siteCoordinates.gridCellCode"                                    
##  [64] "gatheringEvent.siteCoordinates.gridQualifier"                                   
##  [65] "gatheringEvent.siteCoordinates.coordinateErrorDistanceInMeters"                 
##  [66] "gatheringEvent.siteCoordinates.spatialDatum"                                    
##  [67] "gatheringEvent.siteCoordinates.geoShape"                                        
##  [68] "gatheringEvent.namedAreas.areaName"                                             
##  [69] "gatheringEvent.namedAreas.areaClass"                                            
##  [70] "gatheringEvent.associatedTaxa.name"                                             
##  [71] "gatheringEvent.associatedTaxa.relationType"                                     
##  [72] "gatheringEvent.chronoStratigraphy.youngRegionalSubstage"                        
##  [73] "gatheringEvent.chronoStratigraphy.youngRegionalStage"                           
##  [74] "gatheringEvent.chronoStratigraphy.youngRegionalSeries"                          
##  [75] "gatheringEvent.chronoStratigraphy.youngDatingQualifier"                         
##  [76] "gatheringEvent.chronoStratigraphy.youngInternSystem"                            
##  [77] "gatheringEvent.chronoStratigraphy.youngInternSubstage"                          
##  [78] "gatheringEvent.chronoStratigraphy.youngInternStage"                             
##  [79] "gatheringEvent.chronoStratigraphy.youngInternSeries"                            
##  [80] "gatheringEvent.chronoStratigraphy.youngInternErathem"                           
##  [81] "gatheringEvent.chronoStratigraphy.youngInternEonothem"                          
##  [82] "gatheringEvent.chronoStratigraphy.youngChronoName"                              
##  [83] "gatheringEvent.chronoStratigraphy.youngCertainty"                               
##  [84] "gatheringEvent.chronoStratigraphy.oldDatingQualifier"                           
##  [85] "gatheringEvent.chronoStratigraphy.chronoPreferredFlag"                          
##  [86] "gatheringEvent.chronoStratigraphy.oldRegionalSubstage"                          
##  [87] "gatheringEvent.chronoStratigraphy.oldRegionalStage"                             
##  [88] "gatheringEvent.chronoStratigraphy.oldRegionalSeries"                            
##  [89] "gatheringEvent.chronoStratigraphy.oldInternSystem"                              
##  [90] "gatheringEvent.chronoStratigraphy.oldInternSubstage"                            
##  [91] "gatheringEvent.chronoStratigraphy.oldInternStage"                               
##  [92] "gatheringEvent.chronoStratigraphy.oldInternSeries"                              
##  [93] "gatheringEvent.chronoStratigraphy.oldInternErathem"                             
##  [94] "gatheringEvent.chronoStratigraphy.oldInternEonothem"                            
##  [95] "gatheringEvent.chronoStratigraphy.oldChronoName"                                
##  [96] "gatheringEvent.chronoStratigraphy.chronoIdentifier"                             
##  [97] "gatheringEvent.chronoStratigraphy.oldCertainty"                                 
##  [98] "gatheringEvent.bioStratigraphy.youngBioDatingQualifier"                         
##  [99] "gatheringEvent.bioStratigraphy.youngBioName"                                    
## [100] "gatheringEvent.bioStratigraphy.youngFossilZone"                                 
## [101] "gatheringEvent.bioStratigraphy.youngFossilSubZone"                              
## [102] "gatheringEvent.bioStratigraphy.youngBioCertainty"                               
## [103] "gatheringEvent.bioStratigraphy.youngStratType"                                  
## [104] "gatheringEvent.bioStratigraphy.bioDatingQualifier"                              
## [105] "gatheringEvent.bioStratigraphy.bioPreferredFlag"                                
## [106] "gatheringEvent.bioStratigraphy.rangePosition"                                   
## [107] "gatheringEvent.bioStratigraphy.oldBioName"                                      
## [108] "gatheringEvent.bioStratigraphy.bioIdentifier"                                   
## [109] "gatheringEvent.bioStratigraphy.oldFossilzone"                                   
## [110] "gatheringEvent.bioStratigraphy.oldFossilSubzone"                                
## [111] "gatheringEvent.bioStratigraphy.oldBioCertainty"                                 
## [112] "gatheringEvent.bioStratigraphy.oldBioStratType"                                 
## [113] "gatheringEvent.lithoStratigraphy.qualifier"                                     
## [114] "gatheringEvent.lithoStratigraphy.preferredFlag"                                 
## [115] "gatheringEvent.lithoStratigraphy.member2"                                       
## [116] "gatheringEvent.lithoStratigraphy.member"                                        
## [117] "gatheringEvent.lithoStratigraphy.informalName2"                                 
## [118] "gatheringEvent.lithoStratigraphy.informalName"                                  
## [119] "gatheringEvent.lithoStratigraphy.importedName2"                                 
## [120] "gatheringEvent.lithoStratigraphy.importedName1"                                 
## [121] "gatheringEvent.lithoStratigraphy.lithoIdentifier"                               
## [122] "gatheringEvent.lithoStratigraphy.formation2"                                    
## [123] "gatheringEvent.lithoStratigraphy.formationGroup2"                               
## [124] "gatheringEvent.lithoStratigraphy.formationGroup"                                
## [125] "gatheringEvent.lithoStratigraphy.formation"                                     
## [126] "gatheringEvent.lithoStratigraphy.certainty2"                                    
## [127] "gatheringEvent.lithoStratigraphy.certainty"                                     
## [128] "gatheringEvent.lithoStratigraphy.bed2"                                          
## [129] "gatheringEvent.lithoStratigraphy.bed"                                           
## [130] "informationWithheld"                                                            
## [131] "dataGeneralizations"                                                            
## [132] "modified"                                                                       
## [133] "identifications.taxonRank"                                                      
## [134] "identifications.scientificName.fullScientificName"                              
## [135] "identifications.scientificName.taxonomicStatus"                                 
## [136] "identifications.scientificName.genusOrMonomial"                                 
## [137] "identifications.scientificName.subgenus"                                        
## [138] "identifications.scientificName.specificEpithet"                                 
## [139] "identifications.scientificName.infraspecificEpithet"                            
## [140] "identifications.scientificName.infraspecificMarker"                             
## [141] "identifications.scientificName.nameAddendum"                                    
## [142] "identifications.scientificName.authorshipVerbatim"                              
## [143] "identifications.scientificName.author"                                          
## [144] "identifications.scientificName.year"                                            
## [145] "identifications.scientificName.scientificNameGroup"                             
## [146] "identifications.scientificName.references.titleCitation"                        
## [147] "identifications.scientificName.references.citationDetail"                       
## [148] "identifications.scientificName.references.uri"                                  
## [149] "identifications.scientificName.references.author.agentText"                     
## [150] "identifications.scientificName.references.author.fullName"                      
## [151] "identifications.scientificName.references.author.organization.agentText"        
## [152] "identifications.scientificName.references.author.organization.name"             
## [153] "identifications.scientificName.references.publicationDate"                      
## [154] "identifications.scientificName.experts.agentText"                               
## [155] "identifications.scientificName.experts.fullName"                                
## [156] "identifications.scientificName.experts.organization.agentText"                  
## [157] "identifications.scientificName.experts.organization.name"                       
## [158] "identifications.typeStatus"                                                     
## [159] "identifications.dateIdentified"                                                 
## [160] "identifications.defaultClassification.domain"                                   
## [161] "identifications.defaultClassification.subKingdom"                               
## [162] "identifications.defaultClassification.kingdom"                                  
## [163] "identifications.defaultClassification.phylum"                                   
## [164] "identifications.defaultClassification.subPhylum"                                
## [165] "identifications.defaultClassification.superClass"                               
## [166] "identifications.defaultClassification.className"                                
## [167] "identifications.defaultClassification.subClass"                                 
## [168] "identifications.defaultClassification.superOrder"                               
## [169] "identifications.defaultClassification.order"                                    
## [170] "identifications.defaultClassification.subOrder"                                 
## [171] "identifications.defaultClassification.infraOrder"                               
## [172] "identifications.defaultClassification.superFamily"                              
## [173] "identifications.defaultClassification.family"                                   
## [174] "identifications.defaultClassification.subFamily"                                
## [175] "identifications.defaultClassification.tribe"                                    
## [176] "identifications.defaultClassification.subTribe"                                 
## [177] "identifications.defaultClassification.genus"                                    
## [178] "identifications.defaultClassification.subgenus"                                 
## [179] "identifications.defaultClassification.specificEpithet"                          
## [180] "identifications.defaultClassification.infraspecificEpithet"                     
## [181] "identifications.defaultClassification.infraspecificRank"                        
## [182] "identifications.systemClassification.rank"                                      
## [183] "identifications.systemClassification.name"                                      
## [184] "identifications.vernacularNames.name"                                           
## [185] "identifications.vernacularNames.language"                                       
## [186] "identifications.vernacularNames.preferred"                                      
## [187] "identifications.vernacularNames.references.titleCitation"                       
## [188] "identifications.vernacularNames.references.citationDetail"                      
## [189] "identifications.vernacularNames.references.uri"                                 
## [190] "identifications.vernacularNames.references.author.agentText"                    
## [191] "identifications.vernacularNames.references.author.fullName"                     
## [192] "identifications.vernacularNames.references.author.organization.agentText"       
## [193] "identifications.vernacularNames.references.author.organization.name"            
## [194] "identifications.vernacularNames.references.publicationDate"                     
## [195] "identifications.vernacularNames.experts.agentText"                              
## [196] "identifications.vernacularNames.experts.fullName"                               
## [197] "identifications.vernacularNames.experts.organization.agentText"                 
## [198] "identifications.vernacularNames.experts.organization.name"                      
## [199] "identifications.identificationQualifiers"                                       
## [200] "identifications.identifiers.agentText"                                          
## [201] "identifications.taxonomicEnrichments.vernacularNames.name"                      
## [202] "identifications.taxonomicEnrichments.vernacularNames.language"                  
## [203] "identifications.taxonomicEnrichments.synonyms.fullScientificName"               
## [204] "identifications.taxonomicEnrichments.synonyms.taxonomicStatus"                  
## [205] "identifications.taxonomicEnrichments.synonyms.genusOrMonomial"                  
## [206] "identifications.taxonomicEnrichments.synonyms.subgenus"                         
## [207] "identifications.taxonomicEnrichments.synonyms.specificEpithet"                  
## [208] "identifications.taxonomicEnrichments.synonyms.infraspecificEpithet"             
## [209] "identifications.taxonomicEnrichments.synonyms.authorshipVerbatim"               
## [210] "identifications.taxonomicEnrichments.defaultClassification.domain"              
## [211] "identifications.taxonomicEnrichments.defaultClassification.subKingdom"          
## [212] "identifications.taxonomicEnrichments.defaultClassification.kingdom"             
## [213] "identifications.taxonomicEnrichments.defaultClassification.phylum"              
## [214] "identifications.taxonomicEnrichments.defaultClassification.subPhylum"           
## [215] "identifications.taxonomicEnrichments.defaultClassification.superClass"          
## [216] "identifications.taxonomicEnrichments.defaultClassification.className"           
## [217] "identifications.taxonomicEnrichments.defaultClassification.subClass"            
## [218] "identifications.taxonomicEnrichments.defaultClassification.superOrder"          
## [219] "identifications.taxonomicEnrichments.defaultClassification.order"               
## [220] "identifications.taxonomicEnrichments.defaultClassification.subOrder"            
## [221] "identifications.taxonomicEnrichments.defaultClassification.infraOrder"          
## [222] "identifications.taxonomicEnrichments.defaultClassification.superFamily"         
## [223] "identifications.taxonomicEnrichments.defaultClassification.family"              
## [224] "identifications.taxonomicEnrichments.defaultClassification.subFamily"           
## [225] "identifications.taxonomicEnrichments.defaultClassification.tribe"               
## [226] "identifications.taxonomicEnrichments.defaultClassification.subTribe"            
## [227] "identifications.taxonomicEnrichments.defaultClassification.genus"               
## [228] "identifications.taxonomicEnrichments.defaultClassification.subgenus"            
## [229] "identifications.taxonomicEnrichments.defaultClassification.specificEpithet"     
## [230] "identifications.taxonomicEnrichments.defaultClassification.infraspecificEpithet"
## [231] "identifications.taxonomicEnrichments.defaultClassification.infraspecificRank"   
## [232] "identifications.taxonomicEnrichments.sourceSystem.code"                         
## [233] "identifications.taxonomicEnrichments.taxonId"                                   
## [234] "identifications.preferred"                                                      
## [235] "identifications.verificationStatus"                                             
## [236] "identifications.rockType"                                                       
## [237] "identifications.associatedFossilAssemblage"                                     
## [238] "identifications.rockMineralUsage"                                               
## [239] "identifications.associatedMineralName"                                          
## [240] "identifications.remarks"                                                        
## [241] "associatedMultiMediaUris.accessUri"                                             
## [242] "associatedMultiMediaUris.format"                                                
## [243] "associatedMultiMediaUris.variant"                                               
## [244] "theme"</code></pre>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># get info e.g. for field collectionType</span>
<span class="va">sc</span><span class="op">$</span><span class="fu">get_field_info</span><span class="op">(</span><span class="op">)</span><span class="op">$</span><span class="va">content</span><span class="op">$</span><span class="va">collectionType</span></code></pre></div>
<pre><code>## $indexed
## [1] TRUE
## 
## $type
## [1] "keyword"
## 
## $allowedOperators
##  [1] "="                  "!="                 "EQUALS_IC"         
##  [4] "NOT_EQUALS_IC"      "IN"                 "NOT_IN"            
##  [7] "MATCHES"            "NOT_MATCHES"        "STARTS_WITH"       
## [10] "NOT_STARTS_WITH"    "STARTS_WITH_IC"     "NOT_STARTS_WITH_IC"
## 
## $description
## [1] "The subsection of the main Naturalis collection. e.g. Mammals or Fish or Hymenoptera"</code></pre>
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># get all settings</span>
<span class="va">sc</span><span class="op">$</span><span class="fu">get_settings</span><span class="op">(</span><span class="op">)</span></code></pre></div>
<pre><code>## &lt;Response&gt;
## Fields:
##  content: object of class &lt;list&gt; 
##  response: object of class &lt;response&gt; (httr package)</code></pre>
<div class="sourceCode" id="cb31"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># get specific setting</span>
<span class="va">sc</span><span class="op">$</span><span class="fu">get_setting</span><span class="op">(</span><span class="st">"index.max_result_window"</span><span class="op">)</span><span class="op">$</span><span class="va">content</span></code></pre></div>
<pre><code>## [[1]]
## [1] 50000</code></pre>
<div class="sourceCode" id="cb33"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># check if operator is allowed</span>
<span class="va">sc</span><span class="op">$</span><span class="fu">is_operator_allowed</span><span class="op">(</span><span class="st">"gatheringEvent.continent"</span>, <span class="st">"STARTS_WITH"</span><span class="op">)</span><span class="op">$</span><span class="va">content</span></code></pre></div>
<pre><code>## [1] TRUE</code></pre>
<div id="available-paths-and-fields" class="section level4">
<h4 class="hasAnchor">
<a href="#available-paths-and-fields" class="anchor"></a>Available paths and fields</h4>
<p>All fields can be retrieved with <code>get_paths</code> and specific information on the fields, such as allowed operators etc. with <code>get_field_info</code></p>
<div class="sourceCode" id="cb35"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">sc</span><span class="op">$</span><span class="fu">get_paths</span><span class="op">(</span><span class="op">)</span><span class="op">$</span><span class="va">content</span>
<span class="va">sc</span><span class="op">$</span><span class="fu">get_field_info</span><span class="op">(</span><span class="op">)</span><span class="op">$</span><span class="va">content</span></code></pre></div>
</div>
<div id="settings" class="section level4">
<h4 class="hasAnchor">
<a href="#settings" class="anchor"></a>Settings</h4>
<p>Multimedia-specific settings can be retrieved with <code>get_settings</code> and a specific setting with <code>get_setting</code>:</p>
<div class="sourceCode" id="cb36"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">sc</span><span class="op">$</span><span class="fu">get_settings</span><span class="op">(</span><span class="op">)</span><span class="op">$</span><span class="va">content</span></code></pre></div>
<pre><code>## $index.max_result_window
## [1] 50000
## 
## $specimen.group_by_scientific_name.max_num_buckets
## [1] 10000</code></pre>
<div class="sourceCode" id="cb38"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">sc</span><span class="op">$</span><span class="fu">get_setting</span><span class="op">(</span><span class="st">"index.max_result_window"</span><span class="op">)</span><span class="op">$</span><span class="va">content</span></code></pre></div>
<pre><code>## [[1]]
## [1] 50000</code></pre>
</div>
<div id="operators" class="section level4">
<h4 class="hasAnchor">
<a href="#operators" class="anchor"></a>Operators</h4>
<p>To test if a certain operator can be used for a multimedia query:</p>
<div class="sourceCode" id="cb40"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">sc</span><span class="op">$</span><span class="fu">is_operator_allowed</span><span class="op">(</span><span class="st">"identifications.defaultClassification.genus"</span>,
                       <span class="st">"STARTS_WITH"</span><span class="op">)</span><span class="op">$</span><span class="va">content</span></code></pre></div>
<pre><code>## [1] TRUE</code></pre>
</div>
</div>
<div id="dwca-download-services" class="section level3">
<h3 class="hasAnchor">
<a href="#dwca-download-services" class="anchor"></a>DwCA download services</h3>
<p>In addition to query services that return JSON formatted data, the NBA also offers the export of <a href="https://en.wikipedia.org/wiki/Darwin_Core_Archive">Darwin Core Archive (DwCA)</a> files. These files are by default zip archives, please refer to our <a href="https://en.wikipedia.org/wiki/Darwin_Core_Archive">official API documentation</a> for more information.</p>
<div id="static-download" class="section level4">
<h4 class="hasAnchor">
<a href="#static-download" class="anchor"></a>Static download</h4>
<p>Static download services offer the download of predefined datasets. The sets that are available for download can be queried with <code>dwca_get_data_set_names</code>:</p>
<div class="sourceCode" id="cb42"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">sc</span><span class="op">$</span><span class="fu">dwca_get_data_set_names</span><span class="op">(</span><span class="op">)</span><span class="op">$</span><span class="va">content</span></code></pre></div>
<pre><code>##  [1] "amphibia-and-reptilia"      "aves"                      
##  [3] "birdsounds"                 "botany"                    
##  [5] "brachiopoda"                "cainozoic-mollusca"        
##  [7] "chelicerata-and-myriapoda"  "cnidaria"                  
##  [9] "coleoptera"                 "collembola"                
## [11] "crustacea"                  "diptera"                   
## [13] "echinodermata"              "foraminifera"              
## [15] "hemiptera"                  "hymenoptera"               
## [17] "lepidoptera"                "mammalia"                  
## [19] "micropaleontology"          "mollusca"                  
## [21] "observations"               "odonata"                   
## [23] "orthopteroidea"             "paleobotany"               
## [25] "paleontology-invertebrates" "pisces"                    
## [27] "porifera"                   "protista"                  
## [29] "tunicata"</code></pre>
<p>A dataset can then be downloaded using <code>dwca_get_data_set</code>. A filename can be given as argument, if none is given, the DwCA archive is written to <em>download-YYYY-MM-DDThh:mm.zip</em> in the current working directory.</p>
<div class="sourceCode" id="cb44"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># download dataset 'porifera' to temporary file</span>
<span class="va">filename</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/tempfile.html">tempfile</a></span><span class="op">(</span>fileext<span class="op">=</span><span class="st">".zip"</span><span class="op">)</span>
<span class="va">sc</span><span class="op">$</span><span class="fu">dwca_get_data_set</span><span class="op">(</span><span class="st">'porifera'</span>, filename<span class="op">=</span><span class="va">filename</span><span class="op">)</span>                                                                               </code></pre></div>
<pre><code>## Query result written to /tmp/RtmpdxsV1Z/file38e5736b24e.zip</code></pre>
</div>
<div id="dynamic-download" class="section level4">
<h4 class="hasAnchor">
<a href="#dynamic-download" class="anchor"></a>Dynamic download</h4>
<p>The dynamic download function <code>dwca_query</code> allows for download of arbitrary sets, defined by the user’s query. The arguments to this methods are similar to <code>query</code>, plus the filename:</p>
<div class="sourceCode" id="cb46"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># download all specimen of genus 'Hydrochoerus'</span>
<span class="va">filename</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/tempfile.html">tempfile</a></span><span class="op">(</span>fileext <span class="op">=</span> <span class="st">".zip"</span><span class="op">)</span>
<span class="va">qs</span> <span class="op">&lt;-</span>
  <span class="va"><a href="../reference/QuerySpec.html">QuerySpec</a></span><span class="op">$</span><span class="fu">new</span><span class="op">(</span>conditions <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span>
    <span class="va"><a href="../reference/QueryCondition.html">QueryCondition</a></span><span class="op">$</span><span class="fu">new</span><span class="op">(</span>
      field <span class="op">=</span> <span class="st">"identifications.defaultClassification.genus"</span>,
      operator <span class="op">=</span> <span class="st">"EQUALS"</span>,
      value <span class="op">=</span> <span class="st">"Hydrochoerus"</span>
    <span class="op">)</span>
  <span class="op">)</span><span class="op">)</span>
<span class="va">sc</span><span class="op">$</span><span class="fu">dwca_query</span><span class="op">(</span>querySpec <span class="op">=</span> <span class="va">qs</span>, filename <span class="op">=</span> <span class="va">filename</span><span class="op">)</span></code></pre></div>
<pre><code>## Query result written to /tmp/RtmpdxsV1Z/file38e289cf4b4.zip</code></pre>
</div>
</div>
</div>
<div id="taxonomic-data-services" class="section level2 tabset">
<h2 class="hasAnchor">
<a href="#taxonomic-data-services" class="anchor"></a>Taxonomic data services</h2>
<div id="query-services-1" class="section level3">
<h3 class="hasAnchor">
<a href="#query-services-1" class="anchor"></a>Query Services</h3>
<div id="query-1" class="section level4">
<h4 class="hasAnchor">
<a href="#query-1" class="anchor"></a>query</h4>
<p>Query for taxon document with given search criteria. Example:</p>
<div class="sourceCode" id="cb48"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># query for taxa of genus 'Sedum' that are in the Netherlands Soortenregister (NSR)</span>
<span class="va">tc</span> <span class="op">&lt;-</span> <span class="va"><a href="../reference/TaxonClient.html">TaxonClient</a></span><span class="op">$</span><span class="fu">new</span><span class="op">(</span><span class="op">)</span>
<span class="va">qc</span> <span class="op">&lt;-</span>
  <span class="va"><a href="../reference/QueryCondition.html">QueryCondition</a></span><span class="op">$</span><span class="fu">new</span><span class="op">(</span>field <span class="op">=</span> <span class="st">"acceptedName.genusOrMonomial"</span>,
                     operator <span class="op">=</span> <span class="st">"EQUALS"</span>,
                     value <span class="op">=</span> <span class="st">"Sedum"</span><span class="op">)</span>
<span class="va">qc2</span> <span class="op">&lt;-</span>
  <span class="va"><a href="../reference/QueryCondition.html">QueryCondition</a></span><span class="op">$</span><span class="fu">new</span><span class="op">(</span>field <span class="op">=</span> <span class="st">"sourceSystem.code"</span>,
                     operator <span class="op">=</span> <span class="st">"EQUALS"</span>,
                     value <span class="op">=</span> <span class="st">"NSR"</span><span class="op">)</span>
<span class="va">qs</span> <span class="op">&lt;-</span> <span class="va"><a href="../reference/QuerySpec.html">QuerySpec</a></span><span class="op">$</span><span class="fu">new</span><span class="op">(</span>conditions <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span><span class="va">qc</span>, <span class="va">qc2</span><span class="op">)</span><span class="op">)</span>
<span class="va">tc</span><span class="op">$</span><span class="fu">query</span><span class="op">(</span><span class="va">qs</span><span class="op">)</span></code></pre></div>
</div>
<div id="download_query-1" class="section level4">
<h4 class="hasAnchor">
<a href="#download_query-1" class="anchor"></a>download_query</h4>
<p>The <code>query</code> function is limited to retrieve 50000 taxa at once (this is determined in the parameter <code>index.max_result_window</code>, the value is retrievable using the <code>getSettings</code> method in the metadata section). In order to provide access for a larger amount of data, the <code>query_download</code> takes the same arguments as <code>query</code>, but download the data as a gzip stream under the hood. Unlike <code>query</code>, <code>query_download</code> returns a list of taxon objects instead of a <code>ResultSet</code>.</p>
</div>
</div>
<div id="data-access-services-1" class="section level3">
<h3 class="hasAnchor">
<a href="#data-access-services-1" class="anchor"></a>Data access services</h3>
<div id="count-1" class="section level4">
<h4 class="hasAnchor">
<a href="#count-1" class="anchor"></a>count</h4>
<p>For a given query, do not return <code>Taxon</code> objects but the mere count. Example</p>
<div class="sourceCode" id="cb49"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># get counts for taxa of genus 'Sedum' that are in the</span>
<span class="co">#  Netherlands Soortenregister (NSR)</span>
<span class="va">qc</span> <span class="op">&lt;-</span>
  <span class="va"><a href="../reference/QueryCondition.html">QueryCondition</a></span><span class="op">$</span><span class="fu">new</span><span class="op">(</span>field <span class="op">=</span> <span class="st">"acceptedName.genusOrMonomial"</span>,
                     operator <span class="op">=</span> <span class="st">"EQUALS"</span>,
                     value <span class="op">=</span> <span class="st">"Sedum"</span><span class="op">)</span>
<span class="va">qc2</span> <span class="op">&lt;-</span>
  <span class="va"><a href="../reference/QueryCondition.html">QueryCondition</a></span><span class="op">$</span><span class="fu">new</span><span class="op">(</span>field <span class="op">=</span> <span class="st">"sourceSystem.code"</span>,
                     operator <span class="op">=</span> <span class="st">"EQUALS"</span>,
                     value <span class="op">=</span> <span class="st">"NSR"</span><span class="op">)</span>
<span class="va">qs</span> <span class="op">&lt;-</span> <span class="va"><a href="../reference/QuerySpec.html">QuerySpec</a></span><span class="op">$</span><span class="fu">new</span><span class="op">(</span>conditions <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span><span class="va">qc</span>, <span class="va">qc2</span><span class="op">)</span><span class="op">)</span>
<span class="va">tc</span><span class="op">$</span><span class="fu">count</span><span class="op">(</span><span class="va">qs</span><span class="op">)</span></code></pre></div>
<pre><code>## &lt;Response&gt;
## Fields:
##  content: object of class &lt;integer&gt; 
##  response: object of class &lt;response&gt; (httr package)</code></pre>
</div>
<div id="find-1" class="section level4">
<h4 class="hasAnchor">
<a href="#find-1" class="anchor"></a>find</h4>
<p>Returns a taxon object given its identifier:</p>
<div class="sourceCode" id="cb51"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">tc</span><span class="op">$</span><span class="fu">find</span><span class="op">(</span><span class="st">"27706109@COL"</span><span class="op">)</span><span class="op">$</span><span class="va">content</span></code></pre></div>
<pre><code>## Warning in self$handleError(response): 404 (NOT FOUND)
## No Specimen exists with ID 27706109@COL</code></pre>
<pre><code>## [1] "404 (NOT FOUND)\nNo Specimen exists with ID 27706109@COL"</code></pre>
</div>
<div id="find_by_ids-1" class="section level4">
<h4 class="hasAnchor">
<a href="#find_by_ids-1" class="anchor"></a>find_by_ids</h4>
<p>Given a string with comma-separated identifiers, returns a list of taxon objects:</p>
<div class="sourceCode" id="cb54"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">ids</span> <span class="op">&lt;-</span> <span class="st">"27706109@COL,27704140@COL,27706110@COL,27706111@COL,27706108@COL"</span>
<span class="va">res</span> <span class="op">&lt;-</span> <span class="va">tc</span><span class="op">$</span><span class="fu">find_by_ids</span><span class="op">(</span><span class="va">ids</span><span class="op">)</span></code></pre></div>
</div>
</div>
<div id="aggregation-services-1" class="section level3">
<h3 class="hasAnchor">
<a href="#aggregation-services-1" class="anchor"></a>Aggregation services</h3>
<div id="get_distinct_values-1" class="section level4">
<h4 class="hasAnchor">
<a href="#get_distinct_values-1" class="anchor"></a>get_distinct_values</h4>
<p>This method takes a specific field as an argument and returns all possible values and the frequency for that field in the data. Example: get all data source systems for taxon objects:</p>
<div class="sourceCode" id="cb55"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">tc</span><span class="op">$</span><span class="fu">get_distinct_values</span><span class="op">(</span><span class="st">"sourceSystem.name"</span><span class="op">)</span><span class="op">$</span><span class="va">content</span></code></pre></div>
<pre><code>## $`Species 2000 - Catalogue Of Life`
## [1] 1998431
## 
## $`Naturalis - Dutch Species Register`
## [1] 50252
## 
## $`Naturalis - Dutch Caribbean Species Register`
## [1] 8713</code></pre>
</div>
</div>
<div id="metadata-services-1" class="section level3">
<h3 class="hasAnchor">
<a href="#metadata-services-1" class="anchor"></a>Metadata services</h3>
<p>Taxon Metadata services include the same standard metadata services as for the other data types:</p>
<div class="sourceCode" id="cb57"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># get all paths for the Taxon datatype </span>
<span class="va">tc</span><span class="op">$</span><span class="fu">get_paths</span><span class="op">(</span><span class="op">)</span><span class="op">$</span><span class="va">content</span></code></pre></div>
<pre><code>##   [1] "sourceSystem.code"                                       
##   [2] "sourceSystem.name"                                       
##   [3] "sourceSystemId"                                          
##   [4] "recordURI"                                               
##   [5] "sourceSystemParentId"                                    
##   [6] "taxonRank"                                               
##   [7] "taxonRemarks"                                            
##   [8] "occurrenceStatusVerbatim"                                
##   [9] "acceptedName.fullScientificName"                         
##  [10] "acceptedName.taxonomicStatus"                            
##  [11] "acceptedName.genusOrMonomial"                            
##  [12] "acceptedName.subgenus"                                   
##  [13] "acceptedName.specificEpithet"                            
##  [14] "acceptedName.infraspecificEpithet"                       
##  [15] "acceptedName.infraspecificMarker"                        
##  [16] "acceptedName.nameAddendum"                               
##  [17] "acceptedName.authorshipVerbatim"                         
##  [18] "acceptedName.author"                                     
##  [19] "acceptedName.year"                                       
##  [20] "acceptedName.scientificNameGroup"                        
##  [21] "acceptedName.references.titleCitation"                   
##  [22] "acceptedName.references.citationDetail"                  
##  [23] "acceptedName.references.uri"                             
##  [24] "acceptedName.references.author.agentText"                
##  [25] "acceptedName.references.author.fullName"                 
##  [26] "acceptedName.references.author.organization.agentText"   
##  [27] "acceptedName.references.author.organization.name"        
##  [28] "acceptedName.references.publicationDate"                 
##  [29] "acceptedName.experts.agentText"                          
##  [30] "acceptedName.experts.fullName"                           
##  [31] "acceptedName.experts.organization.agentText"             
##  [32] "acceptedName.experts.organization.name"                  
##  [33] "defaultClassification.domain"                            
##  [34] "defaultClassification.subKingdom"                        
##  [35] "defaultClassification.kingdom"                           
##  [36] "defaultClassification.phylum"                            
##  [37] "defaultClassification.subPhylum"                         
##  [38] "defaultClassification.superClass"                        
##  [39] "defaultClassification.className"                         
##  [40] "defaultClassification.subClass"                          
##  [41] "defaultClassification.superOrder"                        
##  [42] "defaultClassification.order"                             
##  [43] "defaultClassification.subOrder"                          
##  [44] "defaultClassification.infraOrder"                        
##  [45] "defaultClassification.superFamily"                       
##  [46] "defaultClassification.family"                            
##  [47] "defaultClassification.subFamily"                         
##  [48] "defaultClassification.tribe"                             
##  [49] "defaultClassification.subTribe"                          
##  [50] "defaultClassification.genus"                             
##  [51] "defaultClassification.subgenus"                          
##  [52] "defaultClassification.specificEpithet"                   
##  [53] "defaultClassification.infraspecificEpithet"              
##  [54] "defaultClassification.infraspecificRank"                 
##  [55] "systemClassification.rank"                               
##  [56] "systemClassification.name"                               
##  [57] "synonyms.fullScientificName"                             
##  [58] "synonyms.taxonomicStatus"                                
##  [59] "synonyms.genusOrMonomial"                                
##  [60] "synonyms.subgenus"                                       
##  [61] "synonyms.specificEpithet"                                
##  [62] "synonyms.infraspecificEpithet"                           
##  [63] "synonyms.infraspecificMarker"                            
##  [64] "synonyms.nameAddendum"                                   
##  [65] "synonyms.authorshipVerbatim"                             
##  [66] "synonyms.author"                                         
##  [67] "synonyms.year"                                           
##  [68] "synonyms.scientificNameGroup"                            
##  [69] "synonyms.references.titleCitation"                       
##  [70] "synonyms.references.citationDetail"                      
##  [71] "synonyms.references.uri"                                 
##  [72] "synonyms.references.author.agentText"                    
##  [73] "synonyms.references.author.fullName"                     
##  [74] "synonyms.references.author.organization.agentText"       
##  [75] "synonyms.references.author.organization.name"            
##  [76] "synonyms.references.publicationDate"                     
##  [77] "synonyms.experts.agentText"                              
##  [78] "synonyms.experts.fullName"                               
##  [79] "synonyms.experts.organization.agentText"                 
##  [80] "synonyms.experts.organization.name"                      
##  [81] "vernacularNames.name"                                    
##  [82] "vernacularNames.language"                                
##  [83] "vernacularNames.preferred"                               
##  [84] "vernacularNames.references.titleCitation"                
##  [85] "vernacularNames.references.citationDetail"               
##  [86] "vernacularNames.references.uri"                          
##  [87] "vernacularNames.references.author.agentText"             
##  [88] "vernacularNames.references.author.fullName"              
##  [89] "vernacularNames.references.author.organization.agentText"
##  [90] "vernacularNames.references.author.organization.name"     
##  [91] "vernacularNames.references.publicationDate"              
##  [92] "vernacularNames.experts.agentText"                       
##  [93] "vernacularNames.experts.fullName"                        
##  [94] "vernacularNames.experts.organization.agentText"          
##  [95] "vernacularNames.experts.organization.name"               
##  [96] "descriptions.description"                                
##  [97] "descriptions.category"                                   
##  [98] "descriptions.language"                                   
##  [99] "descriptions.author"                                     
## [100] "descriptions.license"                                    
## [101] "descriptions.publicationDate"                            
## [102] "references.titleCitation"                                
## [103] "references.citationDetail"                               
## [104] "references.uri"                                          
## [105] "references.author.agentText"                             
## [106] "references.author.fullName"                              
## [107] "references.author.organization.agentText"                
## [108] "references.author.organization.name"                     
## [109] "references.publicationDate"                              
## [110] "experts.agentText"                                       
## [111] "experts.fullName"                                        
## [112] "experts.organization.agentText"                          
## [113] "experts.organization.name"</code></pre>
<div class="sourceCode" id="cb59"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># get info e.g. for field collectionType</span>
<span class="va">tc</span><span class="op">$</span><span class="fu">get_field_info</span><span class="op">(</span><span class="op">)</span><span class="op">$</span><span class="va">content</span><span class="op">$</span><span class="va">synonyms.author</span></code></pre></div>
<pre><code>## $indexed
## [1] TRUE
## 
## $type
## [1] "keyword"
## 
## $allowedOperators
##  [1] "="                  "!="                 "EQUALS_IC"         
##  [4] "NOT_EQUALS_IC"      "CONTAINS"           "NOT_CONTAINS"      
##  [7] "IN"                 "NOT_IN"             "MATCHES"           
## [10] "NOT_MATCHES"        "STARTS_WITH"        "NOT_STARTS_WITH"   
## [13] "STARTS_WITH_IC"     "NOT_STARTS_WITH_IC"</code></pre>
<div class="sourceCode" id="cb61"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># get all settings</span>
<span class="va">tc</span><span class="op">$</span><span class="fu">get_settings</span><span class="op">(</span><span class="op">)</span></code></pre></div>
<pre><code>## &lt;Response&gt;
## Fields:
##  content: object of class &lt;list&gt; 
##  response: object of class &lt;response&gt; (httr package)</code></pre>
<div class="sourceCode" id="cb63"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># get specific setting</span>
<span class="va">tc</span><span class="op">$</span><span class="fu">get_setting</span><span class="op">(</span><span class="st">"index.max_result_window"</span><span class="op">)</span><span class="op">$</span><span class="va">content</span></code></pre></div>
<pre><code>## [[1]]
## [1] 50000</code></pre>
<div class="sourceCode" id="cb65"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># check if operator is allowed</span>
<span class="va">tc</span><span class="op">$</span><span class="fu">is_operator_allowed</span><span class="op">(</span><span class="st">"synonyms.author"</span>, <span class="st">"EQUALS"</span><span class="op">)</span><span class="op">$</span><span class="va">content</span></code></pre></div>
<pre><code>## [1] TRUE</code></pre>
<div id="available-paths-and-fields-1" class="section level4">
<h4 class="hasAnchor">
<a href="#available-paths-and-fields-1" class="anchor"></a>Available paths and fields</h4>
<p>All fields can be retrieved with <code>get_paths</code> and specific information on the fields, such as allowed operators etc. with <code>get_field_info</code></p>
<div class="sourceCode" id="cb67"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">tc</span><span class="op">$</span><span class="fu">get_paths</span><span class="op">(</span><span class="op">)</span><span class="op">$</span><span class="va">content</span>
<span class="va">tc</span><span class="op">$</span><span class="fu">get_field_info</span><span class="op">(</span><span class="op">)</span><span class="op">$</span><span class="va">content</span></code></pre></div>
</div>
<div id="settings-1" class="section level4">
<h4 class="hasAnchor">
<a href="#settings-1" class="anchor"></a>Settings</h4>
<p>Multimedia-specific settings can be retrieved with <code>get_settings</code> and a specific setting with <code>get_setting</code>:</p>
<div class="sourceCode" id="cb68"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">tc</span><span class="op">$</span><span class="fu">get_settings</span><span class="op">(</span><span class="op">)</span><span class="op">$</span><span class="va">content</span></code></pre></div>
<pre><code>## $index.max_result_window
## [1] 50000
## 
## $taxon.group_by_scientific_name.max_num_buckets
## [1] 10000</code></pre>
<div class="sourceCode" id="cb70"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">tc</span><span class="op">$</span><span class="fu">get_setting</span><span class="op">(</span><span class="st">"index.max_result_window"</span><span class="op">)</span><span class="op">$</span><span class="va">content</span></code></pre></div>
<pre><code>## [[1]]
## [1] 50000</code></pre>
</div>
<div id="operators-1" class="section level4">
<h4 class="hasAnchor">
<a href="#operators-1" class="anchor"></a>Operators</h4>
<p>To test if a certain operator can be used for a mutimedia query:</p>
<div class="sourceCode" id="cb72"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">tc</span><span class="op">$</span><span class="fu">is_operator_allowed</span><span class="op">(</span><span class="st">"identifications.defaultClassification.genus"</span>,
                       <span class="st">"STARTS_WITH"</span><span class="op">)</span><span class="op">$</span><span class="va">content</span></code></pre></div>
<pre><code>## Warning in self$handleError(response): Status code:500
## Internal Server Error
## Exception: Invalid element "identifications" in path "identifications.defaultClassification.genus"
## Exception type: nl.naturalis.nba.api.NoSuchFieldException
## Full stack trace stored in response object</code></pre>
<pre><code>## [1] "Internal Server Error"</code></pre>
</div>
</div>
<div id="dwca-download-services-1" class="section level3">
<h3 class="hasAnchor">
<a href="#dwca-download-services-1" class="anchor"></a>DwCA download services</h3>
<p>The taxonomic information in the NBA is also available as Darwin-Core archive files.</p>
<div id="static-download-1" class="section level4">
<h4 class="hasAnchor">
<a href="#static-download-1" class="anchor"></a>Static download</h4>
<p>Static download services offer the download of predefined datasets. The sets that are available for download can be queried with <code>dwca_get_data_set_names</code>:</p>
<div class="sourceCode" id="cb75"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">tc</span><span class="op">$</span><span class="fu">dwca_get_data_set_names</span><span class="op">(</span><span class="op">)</span><span class="op">$</span><span class="va">content</span></code></pre></div>
<pre><code>## [1] "dcsr" "nsr"</code></pre>
<p>A dataset can then be downloaded using <code>dwca_get_data_set</code>. A filename can be given as argument, if none is given, the DwCA archive is written to <em>download-YYYY-MM-DDThh:mm.zip</em> in the current working directory.</p>
<div class="sourceCode" id="cb77"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># download dataset 'nsr' to temporary file</span>
<span class="va">filename</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/tempfile.html">tempfile</a></span><span class="op">(</span>fileext<span class="op">=</span><span class="st">".zip"</span><span class="op">)</span>
<span class="va">tc</span><span class="op">$</span><span class="fu">dwca_get_data_set</span><span class="op">(</span><span class="st">'nsr'</span>, filename<span class="op">=</span><span class="va">filename</span><span class="op">)</span>                                                                               </code></pre></div>
<pre><code>## Query result written to /tmp/RtmpdxsV1Z/file38e42b5925c.zip</code></pre>
</div>
<div id="dynamic-download-1" class="section level4">
<h4 class="hasAnchor">
<a href="#dynamic-download-1" class="anchor"></a>Dynamic download</h4>
<p>The dynamic download function <code>dwca_query</code> allows for download of arbitrary sets, defined by the user’s query. The arguments to this methods are similar to <code>query</code>, plus the filename:</p>
<div class="sourceCode" id="cb79"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># download all taxa for genus 'Clematis'</span>
<span class="va">filename</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/tempfile.html">tempfile</a></span><span class="op">(</span>fileext <span class="op">=</span> <span class="st">".zip"</span><span class="op">)</span>
<span class="va">qs</span> <span class="op">&lt;-</span>
  <span class="va"><a href="../reference/QuerySpec.html">QuerySpec</a></span><span class="op">$</span><span class="fu">new</span><span class="op">(</span>conditions <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span>
    <span class="va"><a href="../reference/QueryCondition.html">QueryCondition</a></span><span class="op">$</span><span class="fu">new</span><span class="op">(</span>
      field <span class="op">=</span> <span class="st">"defaultClassification.genus"</span>,
      operator <span class="op">=</span> <span class="st">"EQUALS"</span>,
      value <span class="op">=</span> <span class="st">"Clematis"</span>
    <span class="op">)</span>
  <span class="op">)</span><span class="op">)</span>
<span class="va">tc</span><span class="op">$</span><span class="fu">dwca_query</span><span class="op">(</span>querySpec <span class="op">=</span> <span class="va">qs</span>, filename <span class="op">=</span> <span class="va">filename</span><span class="op">)</span></code></pre></div>
<pre><code>## Query result written to /tmp/RtmpdxsV1Z/file38e65471923.zip</code></pre>
</div>
</div>
</div>
<div id="geographic-data-services" class="section level2 tabset">
<h2 class="hasAnchor">
<a href="#geographic-data-services" class="anchor"></a>Geographic data services</h2>
<div id="query-services-2" class="section level3">
<h3 class="hasAnchor">
<a href="#query-services-2" class="anchor"></a>Query Services</h3>
<p>The GeoArea query service allows for detailed search within the fields of a GeoArea object. As for the other data types, query parameters can be either given as a <code>list</code> or as a <code>QuerySpec</code> object. Below, we make a simple query to get the <code>GeoArea</code> object for the Netherlands:</p>
<div class="sourceCode" id="cb81"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># instantiate client for geo areas</span>
<span class="va">gc</span> <span class="op">&lt;-</span> <span class="va"><a href="../reference/GeoClient.html">GeoClient</a></span><span class="op">$</span><span class="fu">new</span><span class="op">(</span><span class="op">)</span>

<span class="co"># query for GeoArea of the Netherlands</span>
<span class="va">qc</span> <span class="op">&lt;-</span>
  <span class="va"><a href="../reference/QueryCondition.html">QueryCondition</a></span><span class="op">$</span><span class="fu">new</span><span class="op">(</span>field <span class="op">=</span> <span class="st">"locality"</span>,
                     operator <span class="op">=</span> <span class="st">"EQUALS"</span>,
                     value <span class="op">=</span> <span class="st">"Netherlands"</span><span class="op">)</span>
<span class="va">qs</span> <span class="op">&lt;-</span> <span class="va"><a href="../reference/QuerySpec.html">QuerySpec</a></span><span class="op">$</span><span class="fu">new</span><span class="op">(</span>conditions <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span><span class="va">qc</span><span class="op">)</span><span class="op">)</span>
<span class="va">res</span> <span class="op">&lt;-</span> <span class="va">gc</span><span class="op">$</span><span class="fu">query</span><span class="op">(</span><span class="va">qs</span><span class="op">)</span>

<span class="co"># get item</span>
<span class="va">res</span><span class="op">$</span><span class="va">content</span><span class="op">$</span><span class="va">resultSet</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">$</span><span class="va">item</span></code></pre></div>
<pre><code>## &lt;GeoArea&gt;
## Fields:
##  sourceSystem:   object of class &lt;SourceSystem&gt; 
##  sourceSystemId:  1004050 
##  recordURI:   
##  id:  1004050@GEO 
##  areaType:    Country 
##  locality:    Netherlands 
##  shape:  list of length 2 
##  source:  World Countries 
##  isoCode:     NLD 
##  countryNL:   Nederland 
## Methods:
##  fromJSONString
##  toJSONString
##  fromList
##  toList
##  print</code></pre>
</div>
<div id="data-access-services-2" class="section level3">
<h3 class="hasAnchor">
<a href="#data-access-services-2" class="anchor"></a>Data access services</h3>
<div id="get_geo_json_for_locality" class="section level4">
<h4 class="hasAnchor">
<a href="#get_geo_json_for_locality" class="anchor"></a>get_geo_json_for_locality</h4>
<p>This is a convenience function to directly extract the a GeoJSON object for a specific locality. <a href="https://en.wikipedia.org/wiki/GeoJSON">GeoJSON</a> is a popular format for storing geographical point- and polygon data. To e.g. extract the GeoJSON polygon representation for the Netherlands:</p>
<div class="sourceCode" id="cb83"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">loc</span> <span class="op">&lt;-</span> <span class="st">"Netherlands"</span>
<span class="va">res</span> <span class="op">&lt;-</span> <span class="va">gc</span><span class="op">$</span><span class="fu">get_geo_json_for_locality</span><span class="op">(</span><span class="va">loc</span><span class="op">)</span></code></pre></div>
<p>Results are returned as a <code>list</code> by default, but can be easily converted to a JSON string, e.g.</p>
<div class="sourceCode" id="cb84"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">jsonlite</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/jsonlite/man/fromJSON.html">toJSON</a></span><span class="op">(</span><span class="va">res</span><span class="op">$</span><span class="va">content</span><span class="op">)</span></code></pre></div>
</div>
<div id="count-2" class="section level4">
<h4 class="hasAnchor">
<a href="#count-2" class="anchor"></a>count</h4>
<p>For any given query (with <code>QuerySpec</code> or not), returns the count of matches instead of <code>GeoArea</code> objects:</p>
<div class="sourceCode" id="cb85"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># return count of all GeoAreas</span>
<span class="va">res</span> <span class="op">&lt;-</span> <span class="va">gc</span><span class="op">$</span><span class="fu">count</span><span class="op">(</span><span class="op">)</span>
<span class="va">res</span><span class="op">$</span><span class="va">content</span></code></pre></div>
<pre><code>## [1] 1653</code></pre>
</div>
</div>
<div id="aggregation-services-2" class="section level3">
<h3 class="hasAnchor">
<a href="#aggregation-services-2" class="anchor"></a>Aggregation services</h3>
<div id="get_distinct_values-2" class="section level4">
<h4 class="hasAnchor">
<a href="#get_distinct_values-2" class="anchor"></a>get_distinct_values</h4>
<p>This function returns all values present for a certain field, and their counts:</p>
<div class="sourceCode" id="cb87"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">gc</span><span class="op">$</span><span class="fu">get_distinct_values</span><span class="op">(</span><span class="st">"areaType"</span><span class="op">)</span><span class="op">$</span><span class="va">content</span></code></pre></div>
<pre><code>## $Nature
## [1] 986
## 
## $Municipality
## [1] 408
## 
## $Country
## [1] 247
## 
## $Province
## [1] 12</code></pre>
</div>
</div>
<div id="metadata-services-2" class="section level3">
<h3 class="hasAnchor">
<a href="#metadata-services-2" class="anchor"></a>Metadata services</h3>
<p>Geo Metadata services include the same standard metadata services as for the other data types:</p>
<div class="sourceCode" id="cb89"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># get all paths for the GeoArea datatype </span>
<span class="va">gc</span><span class="op">$</span><span class="fu">get_paths</span><span class="op">(</span><span class="op">)</span><span class="op">$</span><span class="va">content</span></code></pre></div>
<pre><code>##  [1] "sourceSystem.code" "sourceSystem.name" "sourceSystemId"   
##  [4] "recordURI"         "areaType"          "locality"         
##  [7] "shape"             "source"            "isoCode"          
## [10] "countryNL"</code></pre>
<div class="sourceCode" id="cb91"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># get info e.g. for field 'areaType'</span>
<span class="va">gc</span><span class="op">$</span><span class="fu">get_field_info</span><span class="op">(</span><span class="op">)</span><span class="op">$</span><span class="va">content</span><span class="op">$</span><span class="va">areaType</span></code></pre></div>
<pre><code>## $indexed
## [1] TRUE
## 
## $type
## [1] "keyword"
## 
## $allowedOperators
##  [1] "="                  "!="                 "EQUALS_IC"         
##  [4] "NOT_EQUALS_IC"      "IN"                 "NOT_IN"            
##  [7] "MATCHES"            "NOT_MATCHES"        "STARTS_WITH"       
## [10] "NOT_STARTS_WITH"    "STARTS_WITH_IC"     "NOT_STARTS_WITH_IC"</code></pre>
<div class="sourceCode" id="cb93"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># get all settings</span>
<span class="va">gc</span><span class="op">$</span><span class="fu">get_settings</span><span class="op">(</span><span class="op">)</span></code></pre></div>
<pre><code>## &lt;Response&gt;
## Fields:
##  content: object of class &lt;list&gt; 
##  response: object of class &lt;response&gt; (httr package)</code></pre>
<div class="sourceCode" id="cb95"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># get specific setting</span>
<span class="va">gc</span><span class="op">$</span><span class="fu">get_setting</span><span class="op">(</span><span class="st">"index.max_result_window"</span><span class="op">)</span><span class="op">$</span><span class="va">content</span></code></pre></div>
<pre><code>## [[1]]
## [1] 50000</code></pre>
<div class="sourceCode" id="cb97"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># check if operator is allowed</span>
<span class="va">gc</span><span class="op">$</span><span class="fu">is_operator_allowed</span><span class="op">(</span><span class="st">"locality"</span>, <span class="st">"STARTS_WITH"</span><span class="op">)</span><span class="op">$</span><span class="va">content</span></code></pre></div>
<pre><code>## [1] TRUE</code></pre>
</div>
</div>
<div id="multimedia-services" class="section level2 tabset">
<h2 class="hasAnchor">
<a href="#multimedia-services" class="anchor"></a>Multimedia services</h2>
<p>Multimedia services are accessible with a <code>MultimediaClient</code>, instantiated as follows:</p>
<div class="sourceCode" id="cb99"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">mc</span> <span class="op">&lt;-</span> <span class="va"><a href="../reference/MultimediaClient.html">MultimediaClient</a></span><span class="op">$</span><span class="fu">new</span><span class="op">(</span><span class="op">)</span></code></pre></div>
<div id="query-services-3" class="section level3">
<h3 class="hasAnchor">
<a href="#query-services-3" class="anchor"></a>Query Services</h3>
<p>As for the other data types, the <code>query</code> method enables simple and complex queries using a list or a <code>QuerySpec</code> object to specify query parameters.</p>
<div class="sourceCode" id="cb100"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># example of multimedia query passing parameters as a list</span>
<span class="va">mc</span><span class="op">$</span><span class="fu">query</span><span class="op">(</span>queryParams <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span>collectionType <span class="op">=</span> <span class="st">'Cnidaria'</span><span class="op">)</span><span class="op">)</span><span class="op">$</span><span class="va">content</span></code></pre></div>
<pre><code>## &lt;QueryResult&gt;
## Fields:
##  totalSize:   42432 
##  resultSet:  list of length 10 
## Methods:
##  fromJSONString
##  toJSONString
##  fromList
##  toList
##  print</code></pre>
<div class="sourceCode" id="cb102"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># example of multimedia query using QuerySpec: get the first 100</span>
<span class="co"># multimedia items associated with a specimen with name starting with "Ba"</span>
<span class="va">qc</span> <span class="op">&lt;-</span>
  <span class="va"><a href="../reference/QueryCondition.html">QueryCondition</a></span><span class="op">$</span><span class="fu">new</span><span class="op">(</span>field <span class="op">=</span>
                       <span class="st">"identifications.scientificName.fullScientificName"</span>,
                     operator <span class="op">=</span>
                       <span class="st">"STARTS_WITH"</span>,
                     value <span class="op">=</span>
                       <span class="st">"Qu"</span><span class="op">)</span>
<span class="va">qs</span> <span class="op">&lt;-</span> <span class="va"><a href="../reference/QuerySpec.html">QuerySpec</a></span><span class="op">$</span><span class="fu">new</span><span class="op">(</span>conditions <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span><span class="va">qc</span><span class="op">)</span>, size <span class="op">=</span> <span class="fl">100</span><span class="op">)</span>
<span class="va">res</span> <span class="op">&lt;-</span> <span class="va">mc</span><span class="op">$</span><span class="fu">query</span><span class="op">(</span><span class="va">qs</span><span class="op">)</span>

<span class="co"># check if scientific names indeed start with 'Qu'</span>
<span class="fu"><a href="https://rdrr.io/r/base/lapply.html">sapply</a></span><span class="op">(</span><span class="va">res</span><span class="op">$</span><span class="va">content</span><span class="op">$</span><span class="va">resultSet</span>, <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span>
  <span class="va">x</span><span class="op">$</span><span class="va">item</span><span class="op">$</span><span class="va">identifications</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">$</span><span class="va">scientificName</span><span class="op">$</span><span class="va">fullScientificName</span><span class="op">)</span></code></pre></div>
<pre><code>##   [1] "Quercus coccifera L."                                    
##   [2] "Quisqualis indica L."                                    
##   [3] "Quercus spicata Sm."                                     
##   [4] "Quercus sundaica Blume"                                  
##   [5] "Quercus  L."                                             
##   [6] "Quercus heterophylla A.DC."                              
##   [7] "Quercus stellata Wangenh."                               
##   [8] "Quercus hispanica Lam. var. lucombeana (Sweet) Rehder"   
##   [9] "Quercus hispanica Lam. var. lucombeana (Sweet) Rehder"   
##  [10] "Quercus hispanica Lam. var. lucombeana (Sweet) Rehder"   
##  [11] "Quercus hispanica Lam. var. lucombeana (Sweet) Rehder"   
##  [12] "Quercus ×hispanica Lam."                                 
##  [13] "Quercus ×hispanica Lam."                                 
##  [14] "Quercus hispanica Lam."                                  
##  [15] "Quercus hispanica Lam. var. dentata (P.Watson) Rehder"   
##  [16] "Quercus hispanica Lam. var. dentata (P.Watson) Rehder"   
##  [17] "Quercus ×hispanica Lam.  'Fulhamensis'"                  
##  [18] "Quercus havardii Rydb. × gambelii Nutt."                 
##  [19] "Quercus heterophylla A.DC."                              
##  [20] "Quercus heterophylla A.DC."                              
##  [21] "Quercus frainetto Ten."                                  
##  [22] "Quercus frainetto Ten."                                  
##  [23] "Quercus sundaica Blume"                                  
##  [24] "Quercus brandtii Lindl."                                 
##  [25] "Quercus aliena Blume var. acutiserrata Maxim."           
##  [26] "Quercus alba L."                                         
##  [27] "Quercus fernowii Trel."                                  
##  [28] "Quercus coccifera L."                                    
##  [29] "Quercus cerris L. f. argenteovariegata"                  
##  [30] "Quadrella indica (L.) Iltis &amp; Cornejo"                   
##  [31] "Quadrella ferruginea (L.) Iltis &amp; Cornejo"               
##  [32] "Quadrella indica (L.) Iltis &amp; Cornejo"                   
##  [33] "Quiina florida Tul."                                     
##  [34] "Qualea schomburgkiana Warm."                             
##  [35] "Quapoya sipapoana Maguire"                               
##  [36] "Quinchamalium majus Brongn."                             
##  [37] "Quinchamalium raimondii Pilg."                           
##  [38] "Quiina negrensis A.C.Sm."                                
##  [39] "Quiina guianensis Aubl."                                 
##  [40] "Quiina cruegeriana Griseb."                              
##  [41] "Quercus coccinea Münchh."                                
##  [42] "Quercus confusa Wooton &amp; Standl."                        
##  [43] "Quercus confusa Wooton &amp; Standl."                        
##  [44] "Quercus crassifolia Humb. &amp; Bonpl."                      
##  [45] "Quercus coccifera L. var. stenocarpa Pamp."              
##  [46] "Quercus coccinea Münchh."                                
##  [47] "Quercus coccifera L. subsp. calliprinos"                 
##  [48] "Quercus coccifera L."                                    
##  [49] "Quercus coccifera L."                                    
##  [50] "Quercus coccifera L."                                    
##  [51] "Quercus coccifera L."                                    
##  [52] "Quercus coccifera L."                                    
##  [53] "Quercus coccifera L."                                    
##  [54] "Quercus coccifera L."                                    
##  [55] "Quercus coccifera L."                                    
##  [56] "Quercus coccifera L."                                    
##  [57] "Quercus coccifera L."                                    
##  [58] "Quercus coccifera L."                                    
##  [59] "Quercus chrysolepis Liebm."                              
##  [60] "Quercus cerris L. var. austriaca (Willd.) Loud."         
##  [61] "Quercus ×chasei McMinn"                                  
##  [62] "Quercus cerris L."                                       
##  [63] "Quercus cerris L."                                       
##  [64] "Quercus cerris L."                                       
##  [65] "Quercus cerris L."                                       
##  [66] "Quercus cerris L."                                       
##  [67] "Quercus cerris L."                                       
##  [68] "Quercus cerris L."                                       
##  [69] "Quercus cerris L."                                       
##  [70] "Quercus cerris L."                                       
##  [71] "Quercus cerris L."                                       
##  [72] "Quercus castaneifolia C.A.Mey."                          
##  [73] "Quercus castaneifolia C.A.Mey. var. afares (Pomel) Maire"
##  [74] "Quercus canariensis Willd."                              
##  [75] "Quercus canariensis Willd."                              
##  [76] "Quercus brandtii Lindl."                                 
##  [77] "Quercus benderi Baen."                                   
##  [78] "Quercus brandisiana Kurz"                                
##  [79] "Quararibea lasiocalyx (K.Schum.) Vischer"                
##  [80] "Quetzalia occidentalis (Loes. &amp; Donn.Sm.) Lundell"       
##  [81] "Quiina amazonica A.C.Sm."                                
##  [82] "Quiina pteridophylla (Radlk.) Pires"                     
##  [83] "Quiina rhytidopus Tul."                                  
##  [84] "Quercus petraea (Matt.) Liebl."                          
##  [85] "Quercus pubescens Willd."                                
##  [86] "Quercus pubescens Willd."                                
##  [87] "Quercus pubescens Willd."                                
##  [88] "Quercus pubescens Willd. subsp. palensis"                
##  [89] "Quercus macrolepis Kotschy"                              
##  [90] "Quercus petraea (Matt.) Liebl. subsp. iberica"           
##  [91] "Quercus petraea (Matt.) Liebl."                          
##  [92] "Quercus ilex L."                                         
##  [93] "Quercus ilex L."                                         
##  [94] "Quercus ilex L."                                         
##  [95] "Quercus ilex L."                                         
##  [96] "Quercus ilex L."                                         
##  [97] "Quercus gemelliflora Blume"                              
##  [98] "Quercus gemelliflora Blume"                              
##  [99] "Quercus harlandii Hance ex Walp."                        
## [100] "Quercus caudatifolia Merr."</code></pre>
</div>
<div id="data-access-services-3" class="section level3">
<h3 class="hasAnchor">
<a href="#data-access-services-3" class="anchor"></a>Data access services</h3>
<div id="count-3" class="section level4">
<h4 class="hasAnchor">
<a href="#count-3" class="anchor"></a>count</h4>
<p>As for the other data types, a <code>count</code> function returns counts instead of the actual objects:</p>
<div class="sourceCode" id="cb104"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># count all multimedia documents</span>
<span class="va">mc</span><span class="op">$</span><span class="fu">count</span><span class="op">(</span><span class="op">)</span><span class="op">$</span><span class="va">content</span></code></pre></div>
<pre><code>## [1] 14459798</code></pre>
</div>
</div>
<div id="aggregation-services-3" class="section level3">
<h3 class="hasAnchor">
<a href="#aggregation-services-3" class="anchor"></a>Aggregation services</h3>
<div id="get_distinct_values-3" class="section level4">
<h4 class="hasAnchor">
<a href="#get_distinct_values-3" class="anchor"></a>get_distinct_values</h4>
<p>This function returns all values present for a certain field, and their counts. Example: retrieve all different licenses and their counts"</p>
<div class="sourceCode" id="cb106"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">mc</span><span class="op">$</span><span class="fu">get_distinct_values</span><span class="op">(</span><span class="st">"license"</span><span class="op">)</span><span class="op">$</span><span class="va">content</span></code></pre></div>
<pre><code>## $`CC0 1.0`
## [1] 5411392
## 
## $`CC BY-NC-ND 4.0`
## [1] 5048890
## 
## $`All rights reserved`
## [1] 1772832
## 
## $`CC BY-NC-SA 4.0`
## [1] 962879
## 
## $`CC BY-NC 4.0`
## [1] 501421
## 
## $`CC BY 4.0`
## [1] 269074
## 
## $`CC BY-NC-SA 3.0`
## [1] 131598
## 
## $`CC BY-ND 4.0`
## [1] 126687
## 
## $`CC BY-NC-ND 2.5`
## [1] 99512
## 
## $`CC BY-SA 4.0`
## [1] 57502</code></pre>
</div>
</div>
<div id="metadata-services-3" class="section level3">
<h3 class="hasAnchor">
<a href="#metadata-services-3" class="anchor"></a>Metadata services</h3>
<div id="available-paths-and-fields-2" class="section level4">
<h4 class="hasAnchor">
<a href="#available-paths-and-fields-2" class="anchor"></a>Available paths and fields</h4>
<p>All fields can be retrieved with <code>get_paths</code> and specific information on the fields, such as allowed operators etc. with <code>get_field_info</code></p>
<div class="sourceCode" id="cb108"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">mc</span><span class="op">$</span><span class="fu">get_paths</span><span class="op">(</span><span class="op">)</span><span class="op">$</span><span class="va">content</span>
<span class="va">mc</span><span class="op">$</span><span class="fu">get_field_info</span><span class="op">(</span><span class="op">)</span><span class="op">$</span><span class="va">content</span></code></pre></div>
</div>
<div id="settings-2" class="section level4">
<h4 class="hasAnchor">
<a href="#settings-2" class="anchor"></a>Settings</h4>
<p>Multimedia-specific settings can be retrieved with <code>get_settings</code> and a specific setting with <code>get_setting</code>:</p>
<div class="sourceCode" id="cb109"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">mc</span><span class="op">$</span><span class="fu">get_settings</span><span class="op">(</span><span class="op">)</span><span class="op">$</span><span class="va">content</span></code></pre></div>
<pre><code>## $index.max_result_window
## [1] 50000</code></pre>
<div class="sourceCode" id="cb111"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">mc</span><span class="op">$</span><span class="fu">get_setting</span><span class="op">(</span><span class="st">"index.max_result_window"</span><span class="op">)</span><span class="op">$</span><span class="va">content</span></code></pre></div>
<pre><code>## [[1]]
## [1] 50000</code></pre>
</div>
<div id="operators-2" class="section level4">
<h4 class="hasAnchor">
<a href="#operators-2" class="anchor"></a>Operators</h4>
<p>To test if a certain operator can be used for a mutimedia query:</p>
<div class="sourceCode" id="cb113"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">mc</span><span class="op">$</span><span class="fu">is_operator_allowed</span><span class="op">(</span><span class="st">"identifications.scientificName.fullScientificName"</span>,
                       <span class="st">"STARTS_WITH"</span><span class="op">)</span><span class="op">$</span><span class="va">content</span></code></pre></div>
<pre><code>## [1] TRUE</code></pre>
</div>
</div>
</div>
<div id="metadata-services-4" class="section level2 tabset">
<h2 class="hasAnchor">
<a href="#metadata-services-4" class="anchor"></a>Metadata services</h2>
<p>Metadata services provide miscellaneous information about the data available via the NBA. Note that there is also type-specific metadata for each data type (e.g. <code>Specimen</code>) which can be retrieved with the specific client of that class. Here we show the available methods for the <code>MetadataClient</code> which gives general, non-type specific metadata. The client is instantiated in the standard way:</p>
<div class="sourceCode" id="cb115"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">mc</span> <span class="op">&lt;-</span> <span class="va"><a href="../reference/MetadataClient.html">MetadataClient</a></span><span class="op">$</span><span class="fu">new</span><span class="op">(</span><span class="op">)</span></code></pre></div>
<div id="controlled-vocabularies" class="section level4">
<h4 class="hasAnchor">
<a href="#controlled-vocabularies" class="anchor"></a>Controlled vocabularies</h4>
<p>The vocabularies for some fields are controlled by dictionaries with allowed values. For the sex of a museum specimen, for instance, only the terms <em>male</em>, <em>female</em>, <em>mixed</em> and <em>hermaphrodite</em> are allowed to be assigned to the specimen. The fields for which controlled lists are available can be retrieved as follows:</p>
<div class="sourceCode" id="cb116"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">mc</span><span class="op">$</span><span class="fu">get_controlled_lists</span><span class="op">(</span><span class="op">)</span><span class="op">$</span><span class="va">content</span></code></pre></div>
<pre><code>## [1] "AreaClass"          "License"            "LicenseType"       
## [4] "RelationType"       "Sex"                "SpatialDatum"      
## [7] "SpecimenTypeStatus" "TaxonomicStatus"</code></pre>
<p>and for each field that has a controlled vocabulary, there is a separate function to retrieve the allowed values:</p>
<div class="sourceCode" id="cb118"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">mc</span><span class="op">$</span><span class="fu">get_controlled_list_taxonomic_status</span><span class="op">(</span><span class="op">)</span><span class="op">$</span><span class="va">content</span></code></pre></div>
<pre><code>##  [1] "accepted name"               "alternative name"           
##  [3] "ambiguous synonym"           "basionym"                   
##  [5] "homonym"                     "invalid name"               
##  [7] "misapplied name"             "misidentification"          
##  [9] "misspelled name"             "nomen nudum"                
## [11] "preferred name"              "provisionally accepted name"
## [13] "synonym"</code></pre>
<div class="sourceCode" id="cb120"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">mc</span><span class="op">$</span><span class="fu">get_controlled_list_specimen_type_status</span><span class="op">(</span><span class="op">)</span><span class="op">$</span><span class="va">content</span></code></pre></div>
<pre><code>##  [1] "allotype"      "epitype"       "hapantotype"   "holotype"     
##  [5] "isoepitype"    "isolectotype"  "isoneotype"    "isosyntype"   
##  [9] "isotype"       "lectotype"     "neotype"       "paratype"     
## [13] "paralectotype" "syntype"       "topotype"      "type"</code></pre>
<div class="sourceCode" id="cb122"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">mc</span><span class="op">$</span><span class="fu">get_controlled_list_sex</span><span class="op">(</span><span class="op">)</span><span class="op">$</span><span class="va">content</span></code></pre></div>
<pre><code>## [1] "male"          "female"        "mixed"         "hermaphrodite"
## [5] "unknowable"    "undetermined"</code></pre>
<div class="sourceCode" id="cb124"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">mc</span><span class="op">$</span><span class="fu">get_controlled_list_phase_or_stage</span><span class="op">(</span><span class="op">)</span><span class="op">$</span><span class="va">content</span></code></pre></div>
<pre><code>## Warning in self$handleError(response): javax.ws.rs.NotFoundException:
## RESTEASY003210: Could not find resource for full path: http://
## api.biodiversitydata.nl/v2/metadata/getControlledList/PhaseOrStage</code></pre>
<pre><code>## [1] "javax.ws.rs.NotFoundException: RESTEASY003210: Could not find resource for full path: http://api.biodiversitydata.nl/v2/metadata/getControlledList/PhaseOrStage"</code></pre>
</div>
<div id="miscellaneous" class="section level4">
<h4 class="hasAnchor">
<a href="#miscellaneous" class="anchor"></a>Miscellaneous</h4>
<div id="dates" class="section level5">
<h5 class="hasAnchor">
<a href="#dates" class="anchor"></a>Dates</h5>
<p>To maintain data integrity, dates have to be coded in specific formats in our systems. For instance, <em>yyyy-MM-dd</em> is a valid format. Allowed formats can be retrieved as follows:</p>
<div class="sourceCode" id="cb127"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">mc</span><span class="op">$</span><span class="fu">get_allowed_date_formats</span><span class="op">(</span><span class="op">)</span><span class="op">$</span><span class="va">content</span></code></pre></div>
<pre><code>## [1] "yyyy-MM-dd'T'HH:mm:ssZ"         "yyyy-MM-dd'T'HH:mm[:ss][.SSS]Z"
## [3] "yyyy-MM-dd'T'HH:mm[:ss]"        "yyyy-MM-dd HH:mm[:ss]"         
## [5] "yyyy-MM-dd"                     "yyyy-MM"                       
## [7] "yyyy"</code></pre>
</div>
<div id="services-list" class="section level5">
<h5 class="hasAnchor">
<a href="#services-list" class="anchor"></a>Services list</h5>
<p>The method <code>get_rest_services</code> returns a list of all services in the NBA as objects of type <code>RestService</code>.</p>
<div class="sourceCode" id="cb129"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># get the endPoint of the first rest service in the services list</span>
<span class="va">mc</span><span class="op">$</span><span class="fu">get_rest_services</span><span class="op">(</span><span class="op">)</span><span class="op">$</span><span class="va">content</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span></code></pre></div>
<pre><code>## &lt;RestService&gt;
## Fields:
##  endPoint:    / 
##  method:  GET 
##  consumes:    
##  produces:   list of length 1 
##  url:     https://api.biodiversitydata.nl/v2/ 
## Methods:
##  fromJSONString
##  toJSONString
##  fromList
##  toList
##  print</code></pre>
</div>
<div id="settings-3" class="section level5">
<h5 class="hasAnchor">
<a href="#settings-3" class="anchor"></a>Settings</h5>
<p>Similar to the document-specific metadata services, we can get general settings with the <code>MetaDataClient</code>:</p>
<div class="sourceCode" id="cb131"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">#get all settings</span>
<span class="va">mc</span><span class="op">$</span><span class="fu">get_settings</span><span class="op">(</span><span class="op">)</span><span class="op">$</span><span class="va">content</span></code></pre></div>
<pre><code>## $operator.contains.min_term_length
## [1] 3
## 
## $operator.contains.max_term_length
## [1] 15</code></pre>
<div class="sourceCode" id="cb133"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># get value for specific setting</span>
<span class="va">mc</span><span class="op">$</span><span class="fu">get_setting</span><span class="op">(</span><span class="st">"operator.contains.min_term_length"</span><span class="op">)</span><span class="op">$</span><span class="va">content</span></code></pre></div>
<pre><code>## [[1]]
## [1] 3</code></pre>
</div>
</div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><!-- begin footer --><div class="footer">
    <div class="container">
        <div class="row start bottom-8">
            <div class="col-xs-2"> <img id="footerlogo" src="https://ropensci.org/img/icon_short_white.svg">
</div>
            <div class="col-xs-10">
                <div class="row">
                    <div class="col-md-4 col-xs-6">
                        <a href="http://github.com/ropensci" target="_blank"><div class="icon fab fa-github"></div></a>
                        <a href="http://github.com/ropenscilabs" target="_blank"><div class="icon fa fa-flask"></div></a>
                        <a href="http://twitter.com/ropensci" target="_blank"><div class="icon fab fa-twitter"></div></a>
                        <a href="http://vimeo.com/ropensci" target="_blank"><div class="icon fab fa-vimeo"></div></a>
                    </div>
                </div>
                <div class="row top-4">
                    <div class="col-md-2 col-sm-4">
                        <ul>
<h5 class="bottom-2">About</h5>
                            <li><a href="https://ropensci.org/about">About rOpenSci</a></li>
                            <li><a href="https://ropensci.org/software-review">Software Review Process</a></li>
                            <li><a href="https://ropensci.org/about#team">Our Team</a></li>
                            <li><a href="https://ropensci.org/careers">Jobs</a></li>
                            <li><a href="https://ropensci.org/donate">Donate</a></li>
                            <li><a href="https://ropensci.org/contact">Contact Us</a></li>
                        </ul>
</div>
                    <div class="col-md-3 col-sm-4">
                        <ul>
<h5 class="bottom-2">Community</h5>
                            <li><a href="https://ropensci.org/community/">Our Community</a></li>
                            <li><a href="https://ropensci.org/commcalls/">Community calls</a></li>
                            <li><a href="https://ropensci.org/events/">Events</a></li>
                            <li><a href="https://discuss.ropensci.org/">Join the Discussion</a></li>
                            <li><a href="https://ropensci.org/code-of-conduct">Code of conduct</a></li>
                        </ul>
</div>
                    <div class="col-md-2 col-sm-4">
                        <ul>
<h5 class="bottom-2">Resources</h5>
                            <li><a href="https://ropensci.org/packages/">Packages</a></li>
                            <li><a href="https://ropensci.org/usecases/">Use Cases</a></li>
                            <li><a href="https://ropensci.org/talks-papers/">Talks &amp; Publications</a></li>
                            <li><a href="https://docs.ropensci.org/">Documentation</a></li>
                            <li><a href="https://ropensci.org/how-to-cite-ropensci/">Cite rOpenSci</a></li>
                        </ul>
</div>
                    <div class="col-md-4 col-xs-12">
                        <h5 class="bottom-2"></h5>

                        <p>rOpenSci is a fiscally sponsored project of <a href="http://numfocus.org" target="_blank">NumFOCUS</a></p>
                        <a href="http://numfocus.org" target="_blank"></a>

                    </div>
                </div>
            </div>
        </div>
    </div>
</div>
<!-- / end footer -->

      </footer>
</div>

  
<script src="https://cdnjs.cloudflare.com/ajax/libs/docsearch.js/2.6.1/docsearch.min.js" integrity="sha256-GKvGqXDznoRYHCwKXGnuchvKSwmx9SRMrZOTh2g4Sb0=" crossorigin="anonymous"></script><script>
  docsearch({
    
    
    apiKey: '799829e946e1f0f9cd5b5a782c6316b9',
    indexName: 'ropensci-nbar',
    inputSelector: 'input#search-input.form-control',
    transformData: function(hits) {
      return hits.map(function (hit) {
        hit.url = updateHitURL(hit);
        return hit;
      });
    }
  });
</script>
</body>
</html>
